flake.nix

```diff
@@ -1,55 +1,73 @@
 {
   description = "flora";
   inputs = {
-    flake-utils = { url = "github:numtide/flake-utils"; };
-    horizon-platform = {
-      url = "git+https://gitlab.horizon-haskell.net/package-sets/horizon-platform";
-    };
+    flake-utils.url = "github:numtide/flake-utils";
+    horizon-platform.url =
+      "git+https://gitlab.horizon-haskell.net/package-sets/horizon-platform";
     nixpkgs.url = "github:nixos/nixpkgs/nixpkgs-unstable";
-    streamly.url = "git+https://github.com/composewell/streamly";
-    streamly.flake = false;
-    poolboy.url = "git+https://github.com/blackheaven/poolboy";
+    # nixos-22.05 provides souffle 2.2; that version does not easily
+    # build on the current pin, so use a separate input for this
+    nixpkgs-stable.url = "github:nixos/nixpkgs/nixos-22.05";
+    streamly = {
+      url = "github:composewell/streamly";
+      flake = false;
+    };
+    poolboy.url = "github:blackheaven/poolboy";
   };
-  outputs =
-    inputs@{ self, flake-utils, horizon-platform, nixpkgs, ... }:
+  outputs = inputs@{ self, flake-utils, horizon-platform, nixpkgs, ... }:
     flake-utils.lib.eachSystem [ "x86_64-linux" ] (system:
-    let
-      pkgs = import nixpkgs { inherit system; };
-      hsPkgs = with pkgs.haskell.lib;
-        horizon-platform.legacyPackages.${system}.override {
-          overrides = hfinal: hprev: rec {
-            flora = overrideCabal
-              (dontHaddock (dontCheck
-                (doJailbreak (hfinal.callCabal2nix "flora" ./. { }))))
-              (drv: {
-                preConfigure = ''
-                  cd cbits; ${pkgs.souffle}/bin/souffle -g categorise.{cpp,dl}
-                  cd ..
-                '';
-              });
-            streamly-core = hfinal.callCabal2nix "streamly-core" "${inputs.streamly}/core/" { };
-            streamly = hfinal.callCabal2nix "streamly" inputs.streamly { };
-            poolboy = hfinal.callCabal2nix "poolboy" inputs.poolboy { };
+      let
+        pkgs = import nixpkgs { inherit system; };
+        souffle_2_2 = inputs.nixpkgs-stable.legacyPackages.${system}.souffle;
+        hsPkgs = with pkgs.haskell.lib;
+          horizon-platform.legacyPackages.${system}.override {
+            overrides = hfinal: hprev: {
+              flora = overrideCabal (dontHaddock (dontCheck
+                (doJailbreak (hfinal.callCabal2nix "flora" ./. { })))) (drv: {
+                  preConfigure = ''
+                    cd cbits
+                    ${pkgs.lib.getExe souffle_2_2} -g categorise.{cpp,dl}
+                    cd ..
+                  '';
+                });
+              streamly-core =
+                hfinal.callCabal2nix "streamly-core" "${inputs.streamly}/core/"
+                { };
+              streamly = hfinal.callCabal2nix "streamly" inputs.streamly { };
+              poolboy = hfinal.callCabal2nix "poolboy" inputs.poolboy { };
```

Comment:
this needs `dontCheck`, does this work for you @kenranunderscore ? I don't think so tbh.

Reply:
I've tried picking this up myself and this doesn't build because of a test failure in `poolboy`

Reply:
I will fix this.

Reply:
i also just saw that this uses `horizon-haskell` in that case we shouldn't add nixpkgs 22.11 to the inputs, it should just use callHackage.

Reply:
I can't follow :sweat_smile: sorry. I can enter the `nix develop` environment just fine, but as I mentioned in the original post I don't use the provided packages, but rather let `cabal` download the frozen versions.

What are you doing/executing that leads to the failure?
(It's not 22.11 btw, but 22.05, as that (and I think only that) has the correct `souffle` version.)

Reply:
Ah, I think I see what you mean now though with `pkgs`. I'll have to go out for a bit; let me try and check whether just dropping one of the `nixpkgs` inputs just works.

Reply:
`poolboy` tests fail on my machine with your code. I guess it has flaky tests? I have just added dontCheck. I will open a PR as soon as I have this done

Reply:
i kinda wonder whether we should just use something else but horizon-haskell, as nix will not be the source of trug

Reply:
I'm absolutely open to your opinion on the boundaries of nix and cabal in this project.

Reply:
so the issue I see here is that given that you have a lot of custom pinned dependencies in your `cabal.project`, without a maintainer for the nix code, they will probably get out of sync with the deps pinned, there are multiple solutions: 
- use IFD for parsing a lock file from this repo to nix code, so the nix generated will adapt to version changes and the maintenance burden is reduce (e.g. by using something like stack lock or cabal freeze file) 
- check the nix in the actions and call a maintainer if it breaks

Reply:
then we also have issue like the above where souffle 2.3 is on nixpkgs-stable since november '22 and flora doesn't keep up which cannot be solved in general, we can keep pins but if flora is advancing in one part of their dependencies and not in the other parts, then it's hard to provide consistent package sets

Reply:
There's also a "middle way", which is what I'm using locally right now, of just providing the tooling needed to develop the application, as version bounds don't matter as much there and usually there's just a minimal bound.

In its most extreme form, that would mean using nix to just provide `cabal-install`, `ghc`, etc. and also stuff like `souffle`, `postgresql-migration`, `nixfmt`, etc. in a development environment, where you'd then use the cabal freeze for "normal" non-nixy development.

I actually like this approach (for now), as:
1. `nix build` doesn't work anyway, and never did.
2. The `flora` package in the package set needs to be jailbroken, tests disabled and so on, and is ultimately only used to provide its dependencies in the nix shell; but these dependencies don't work atm either, and are different from the frozen ones.
3. It would simplify the flake *a lot*, that is, the "migration path" to this kind of setup is basically: remove the non-working parts and transitive inputs, then flesh out the `devShell`.

But should we make an issue to discuss this further maybe?

I see value in the PR as-is (potentially with @MangoIV's `dontCheck`), but could also convert it into the above proposal if you like.

Reply:
My priority right now is to enable my nix-using contributors to get started with minimal amounts of nerd-sniping. I trust your judgement on what to do to achieve this goal. :)